### PR TITLE
Add UI option to show uncertainties in specviz viewer

### LINKED ISSUE #285: Mask and uncertainty plotting only works for first spectrum
Currently, the Specviz uncertainty and mask plotting is only available for the first spectrum loaded into the viewer. We should refactor/enhance this so that all/only displayed spectra have their uncertainties and masks plotted when those methods are called.

### LINKED ISSUE #1146: [BUG] Specviz: uncertainty and mask unavailable for load_spectrum loaded spectra
I just did the following (don't worry about the details of how I loaded the spectrum from a fits file, I'm just using that to illustrate the structure of the Spectrum1D file):
```
spec = specutils.Spectrum1D(flux=data[1].data[fiberrow]*u.count, 
                     spectral_axis=data[0].data[fiberrow]*u.angstrom,
                     uncertainty=InverseVariance(data[2].data[233]),
                     mask=data[3].data[fiberrow])
viz = Specviz()
viz.load_spectrum(spec)
viz.app
```
And the result was a perfectly fine-looking spectrum, but the uncertainty and the mask were nowhere to be found in specviz.

More curiously, when I did `viz.get_spectra()`, the resulting spectrum had the right `spectral_axis`/`uncertainty`/`flux`, but the `mask` was absent! 

This *may* be related to #285 ?

### PR DESCRIPTION
**Pull request by javerbukh**
### Description

This pull request is to allow the user to show the uncertainty of a spectrum in the specviz viewer using a UI toggle. It also fixes the code for showing uncertainties and masks, allowing them to work for multiple data selected in the specviz viewer.

Fixes #1146 fixes #285

Screen Shot 2022-03-22 at 12 38 23 PM

### TODO

- [ ] Make test coverage job pass.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (javerbukh) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:60]: I know we want to move towards using `app.get_viewer_by_id()` but this line works with both specviz and cubeviz. `app.get_viewer_by_id()` would require the `id` to be `specviz-0` or `cubeviz-3`, respectively.

**Comment by PatrickOgle** [comment]
I like how the continuous/shaded errorbars look.  Will there also be an option to plot separate y errorbars on each data point?

**Comment by rosteen** [comment]
Woah, 285 is a low number these days! Nice fix, it works in my initial testing but I'll try a little harder to break it and read through the code changes before I approve.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/viewers.py:313]: should we just hide the marks instead of deleting them?  (I'm honestly not sure the performance implications and this isn't something that will be toggled repeatedly, so maybe its not worth it)

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (rosteen) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:50]: Is there a reason you had to default to turning off the uncertainties when new data is added, rather than showing them for the new data as well if they were toggled on?

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/viewers.py:379]: this is likely catching subsets as well, which is causing a shape-mismatch error when creating a subset and then enabling uncertainties in the plot options.  And probably also won't trigger when new data is added unless the uncertainty marks are all cleared and then recreated.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/viewers.py:394]: ok, that could be handy!  Is this controlled by the same switch as uncertainties or should they be independent?

**Comment by javerbukh** [comment]
@PatrickOgle That was not in the acceptance criteria, so I had no plan to implement that. That sounds like it could be a continuation ticket for this (i.e. add more UI options, add more visualization options, etc.)

@rosteen Haha, it is quite old! I searched our open issues for anything related to this and that came up.

@kecnry and Ricky, thank you for the comments! I'll work on the next commit for this and have it up soon.

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:50]: Good point. I made this change before I went in and refactored the uncertainties/mask code on the backend. I can switch this to work as you described.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (javerbukh) [reply jdaviz/configs/specviz/plugins/viewers.py:313]: I actually didn't know we could hide marks instead of deleting them...I'll look into that!

Comment (rosteen) [reply jdaviz/configs/specviz/plugins/viewers.py:313]: I _think_ every mark has a `.visible` property that can be toggled, iirc.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (rosteen) [reply jdaviz/configs/specviz/plugins/viewers.py:379]: Probably something to do with the `subset_state` attribute, but I don't have time right this moment to look into it. I think that's attached to layers but I'm not 100% sure. Maybe search for that in jdaviz/glue-astronomy.

Comment (javerbukh) [reply jdaviz/configs/specviz/plugins/viewers.py:379]: Right! I think that may be the answer, thanks!

File: jdaviz/core/marks.py
Comment (kecnry) [inline jdaviz/core/marks.py:269]: If we split out some of the logic from `BaseSpectrumVerticalLine` into an even-higher base class, then these could inherit from that and get automatic unit conversion support (although we'll need to write the y-unit logic yet).

Alternatively, these could be implemented similarly to the `Shadow` mark class and avoid the need for unit conversion entirely (by just being plotted relative to the parent spectrum object).

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:60]: This doesn't exist for imviz, for example, but the toggle uncertainty button still shows (and actually causes imviz to crash).

**Comment by javerbukh** [comment]
@rosteen Its related to something else I found while testing which is that if you change the color of a spectrum (added after the first one), then de-select it, and turn uncertainties off and on, the shading changes to the color. I believe that error is here, but that would mean this PR just exposed it and did not create it:
https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/specviz/plugins/viewers.py#L388-L393

**Comment by rosteen** [comment]
Interesting, I do see the same thing following that workflow. I tried confirming that it's present on main as well, but the old implementation just breaks trying to do the same thing. So this is at least an improvement 😆 

I think this is useful enough as-is that I'd be ok opening a follow-up issue to investigate the layer color weirdness.

**Comment by rosteen** [comment]
I'm going to take 20 minutes to see if I can come up with a quick fix, if I can't I'll approve this.

File: jdaviz/configs/default/plugins/plot_options/plot_options.vue
Comment (kecnry) [inline jdaviz/configs/default/plugins/plot_options/plot_options.vue:23]: `_toggle_uncertainty` currently handles specviz and cubeviz, but this says specviz and mosviz.  Also see comment above about spectrum viewers in specviz2d and mosviz.

**Comment by javerbukh** [comment]
@rosteen I fixed the bug and got rid of the `data_trace_pointer` I'll work on @kecnry 's suggestions now.

File: jdaviz/configs/default/plugins/plot_options/plot_options.vue
Comment (javerbukh) [reply jdaviz/configs/default/plugins/plot_options/plot_options.vue:23]: So if I only want it to not appear in Imviz, do I place 'imviz' in the brackets and set `.indexOf(config)!==1`? Or leave it as `-1`?

**Comment by rosteen** [comment]
> @rosteen I fixed the bug and got rid of the `data_trace_pointer` I'll work on @kecnry 's suggestions now.

Ha, I was in the middle of testing exactly this, was going to submit a commit. Looks good to me.

**Comment by rosteen** [comment]
Bollocks, I hit "close with comment" whoops.

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (javerbukh) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py]: Fixed.

**Comment by javerbukh** [comment]
@rosteen @kecnry I think I've addressed all of your review comments (so far). Let me know if anything else needs work!

File: jdaviz/configs/default/plugins/plot_options/plot_options.vue
Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.vue]: yes, -1 is returned if the item isn't in the list, so `list.indexOf(item)!==-1` is equivalent to python's `item in list`

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/viewers.py:394]: does removing this have consequences in any other situation that might not be covered by CI?

**Review by rosteen — APPROVED** [review]
Looks good to me now, thanks for addressing those comments!

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py:394]: Searched the whole code base and these attributes don't seem to appear elsewhere, so hopefully its safe 🤞

**Comment by pllim** [comment]
Well, already has 2 approvals, so :shipit: ❗ 

Thanks, everyone!

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: Actually, will such hardcoding break MAST? cc @havok2063

File: jdaviz/core/marks.py
Comment (pllim) [reply jdaviz/core/marks.py:292]: Please link here if you do decide to create a follow-up ticket. Thanks!

File: jdaviz/configs/default/plugins/plot_options/plot_options.py
Comment (pllim) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: To debate with myself, I don't think it'll break MAST though we might need extra logic if MAST wants this feature. Future work!

Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: Do you mean the per-config hardcoded viewer IDs?

Comment (havok2063) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: I don't see where the hard-coding is taking place?  Lines 60-66 below? But my answer without context is "it might break mast??".  Currently in our notebook we customize the `settings.configuration` value and call it "mastviz" when we pass it into the main application.  It sounds like that might not be a good idea if internal jdaviz code checks for one of the specific viztools for functionality.    We currently customize config items from `settings`, `toolbar`, and `tray`, but not `viewer_area`.

Comment (havok2063) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: There's no other reason for us renaming the configuration except to explicitly say "this is not exactly the same Cubeviz, Specviz, etc that you'll see in a notebook".  And of course we only say this to people who look at the internal notebook, which is me.

Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: It might be best to just loop over all viewers in the app and call `show_uncertainties` on any that are an instance of `SpecvizProfileView`.   The UI logic also needs to be changed from checking the app-config to checking the type of the currently selected viewer (so the switch does not show in the image viewers in cubeviz, etc)

Comment (havok2063) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: I will also change our notebook so it keeps the original configuration name.

Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:59]: see #1208 for changing this to per-viewer and removing reliance on the configuration name.

**Pull request merged**
